### ISSUE TITLE
clipboard.get_selection() when no text is selected.

### ISSUE THREAD

**Issue by vvasuki** [autokey/autokey#553]
## Classification:

Bug or Enhancement

## Reproducibility:

Always

## Version

AutoKey version: 0.95.10

Used GUI (Gtk, Qt, or both): Gtk

Installed via: (PPA, pip3, …). yay/ pacman

Linux Distribution: Arch linux

## Summary
`clipboard.get_selection()` should return empty string when no text is selected.

## Steps to Reproduce (if applicable)

- Copy some text.
- Don't select any text.
- Invoke some hotkey command to access the current selection via `clipboard.get_selection`

## Expected Results

You should get an empty string or an exception

## Actual Results

You get whatever is in the clipboard.

**Comment by josephj11** [comment]
TL;DR: It's not a bug, it's a feature (of X11). (Or, if it is a bug conflating the two clipboards, it's way upstream of us.)

Between not many people actually understanding the clipboard architecture and various clipboard managers that may alter the clipboard behavior, you can't really rely on things working exactly as you think they should.

The rest of this is probably not relevant to this issue, but I wrote it before I figured that out and someone else may find it useful in the future. :)

My very imperfect understanding of Linux clipboards are that there are 3 of them. The clipboard, the primary selection and a secondary selection (that almost no one uses).

Clipboards seem really simple, but there's a lot [under the hood](https://www.jwz.org/doc/x-cut-and-paste.html). 

From what I read, when you select some text the program you use asserts ownership of the selection, but doesn't do anything else. When you try to copy the selection, it is requested from that owner. If the owner is still running, you get it.

The various documents I have read don't say anything about these clipboards getting reset or cleared unless the selection owner stops running or you clear it explicitly.

For our purposes, if you ever selected something since you logged in, the program will own the selection and be ready to provide it in the indefinite future as long as it's still running. This is independent of anything being selected at the exact moment you ask for it. 

In fact, I believe you do get an exception if the clipboard you're accessing is really empty. See #236

**Comment by vvasuki** [comment]
Thanks for the clarification.

So, how would I rewrite the below so as to determine if the user has selected something at the moment the Autokey script is invoked (and not in the past)?

```
text = clipboard.get_selection()
if text == "":
    output = "-"
elif text[0] >= "क" and  text[0] <= "ॿ" :
    output = text[0]+"् "
```

**Comment by ineuw** [comment]
Try this to clear the clipboard.

import time
clipboard.fill_clipboard("")
time.sleep(0.1) # or greater 

"clipboard" in Autokey always refers to the standard "Clipboard" which handles copy and paste. The other two X "clipboards" are not involved, unless they are referred to specifically.

References:-the ICCCM, obviously-https://www.jwz.org/doc/x-cut-and-paste.html

**Comment by josephj11** [comment]
@vvasuki My first guess is that there's no way to do that unless you somehow first clear the clipboard as @ineuw suggests *before* this script is launched.

However, since this question is really just about X11 clipboard behavior, independent of AutoKey or any other particular  application, you could make a general  question and ask about it on stackoverflow where lots of Linux experts live.

If you do that and get and get any further clarification, please let us know what you found out.

If this functionality is critical to your application, you could get freaky and do things like having another app (not in AutoKey) running which periodically keeps reading the clipboard and setting a timestamp whenever it changes. It could store that timestamp somewhere (e.g. by touching or editing a file) where your AutoKey script could retrieve it and know how old it is. That would give you a rough idea of whether the data is current or old.

If that's way too convoluted for you, then just make sure you give your user some way to back out of whatever your script does if it gets old data.

**Comment by ineuw** [comment]
I copied and kept the original because it was so important. That web page no longer exists, but here is contents: 

References:-the ICCCM, obviously-http://www.xfree86.org/~keithp/talks/selection.ps-http://www.jwz.org/doc/x-cut-and-paste.html 
PRIMARY - This is normally used for middle mouse button = copy/paste
SECONDARY - Normally not used by much, but it exists. Usually as an in app specific copy and paste.
CLIPBOARD - Usually Ctrl+c and Ctrl+v style copy and paste.

clients have not handled cut-and-paste in a consistent way, leading users to believe that X doesn't even have a working clipboard. This isn't really accurate; there is a conventional behavior, somewhat standardized in the ICCCM. But many apps don't follow the conventional behavior.

X has a thing called "selections" which are just clipboards, essentially (implemented with an asynchronous protocol so you don't actually copy data to them). There are three standard selections defined in the ICCCM: PRIMARY, SECONDARY, and CLIPBOARD.

The ICCCM defines these as follows:
"The selection named by the atom PRIMARY is used for all commands that take only a single argument and is the principal means of communication between clients that use the selection mechanism.

The selection named by the atom SECONDARY is used:
o As the second argument to commands taking two arguments (for example, "exchange primary and secondary selections")

o As a means of obtaining data when there is a primary selection and the user does not want to disturb it

The selection named by the atom CLIPBOARD is used to hold data that is being transferred between clients, that is, data that usually is being cut and then pasted or copied and then pasted."

In addition, the ICCCM has a thing called "cut buffers" which most clients no longer support.

There are two historical interpretations of the ICCCM:

a) use PRIMARY for mouse selection, middle mouse button paste, and explicit cut/copy/paste menu items (Qt 2, GNU Emacs 20) b) use CLIPBOARD for the Windows-style cut/copy/paste menu items; use PRIMARY for the currently-selected text, even if it isn't explicitly copied, and for middle-mouse-click (Netscape, Mozilla, XEmacs, some GTK+ apps)

No one ever does anything interesting with SECONDARY as far as I can tell.

The current consensus is that interpretation b) is correct. Qt 3 and GNU Emacs 21 will use interpretation b), changing the behavior of previous versions.

The correct behavior can be summarized as follows: CLIPBOARD works just like the clipboard on Mac or Windows; it only changes on explicit cut/copy. PRIMARY is an "easter egg" for expert users, regular users can just ignore it; it's normally pastable only via middle-mouse-click.

The rationale for this behavior is mostly that behavior a) has a lot of problems, namely:

-inconsistent with Mac/Windows-confusingly, selecting anything overwrites the clipboard-not efficient with a tool such as xclipboard-you should be able to select text, then paste the clipboard over it, but that doesn't work if the selection and clipboard are the same-the Copy menu item is useless and does nothing, which is confusing-if you think of PRIMARY as the current selection, Cut doesn't make any sense since the selection simultaneously disappears and becomes the current selection

Application authors should follow the following guidelines to get correct behavior:

— selecting but with no explicit copy should only set PRIMARY, never CLIPBOARD

— middle mouse button should paste PRIMARY, never CLIPBOARD

— explicit cut/copy commands (i.e. menu items, toolbar buttons) should always set CLIPBOARD to the currently-selected data (i.e. conceptually copy PRIMARY to CLIPBOARD)

— explicit cut/copy commands should always set both CLIPBOARD and PRIMARY, even when copying doesn't involve a selection (e.g. a "copy url"-option which explicitly copies an url without the url being selected first)

— explicit paste commands should paste CLIPBOARD, not PRIMARY

— a selection becoming unselected should never unset PRIMARY

— possibly contradicting the ICCCM, clients don't need to support SECONDARY, though if anyone can figure out what it's good for they should feel free to use it for that

— cut buffers are evil; they only support ASCII, they don't work with many clients, and they require data to be copied to the X server. Therefore clients should avoid using cut buffers and use only selections.

Apps that follow these guidelines give users a simple mental model to understand what's going on. PRIMARY is the current selection. Middle button pastes the current selection. CLIPBOARD is just like on Mac/Windows. You don't have to know about PRIMARY if you're a newbie.

I don't think there's a sane mental model if we collapse everything into PRIMARY and ignore clipboard, see rationale above.

A remaining somewhat odd thing about X selections is that exiting the app you did a cut/copy from removes the cut/copied data from the clipboard, since the selection protocol is asynchronous and requires the source app to provide the data at paste time. The solution here would be a standardized protocol for a "clipboard daemon" so that apps could hand off their data to a daemon when they exit. Or alternatively, you can run an application such as xclipboard which constantly "harvests" clipboard selections.

**Comment by ineuw** [comment]
I apologize for length of my previous post, but there is more. Another solution is to install "xsel" and call it in your script to clear the clipboard with "xsel -bc" or use  "xsel --help" for additional options. Regardless of the selected solution, time.sleep() is required. In my case, after numerous experimentation resulted in the time.sleep(0.2) to clear the clipboard of 700~ words. Less than that, old info was retained to various extent.

**Comment by BlueDrink9** [comment]
@josephj11 since we are getting buried under issues that are often support requests or unrelated bugs, I think it would be good to close issues more aggressively. Then the opener can argue for them to be re-opened, if they think there is something that devs can actually do about it.

In this case, unless there is something autokey can change, this should be closed

**Comment by vvasuki** [comment]
Thanks for all the tips, everyone. 

It would be splendid if autokey itself could do this (without needing the user to write another app - being a tall order) proposed in https://github.com/autokey/autokey/issues/553#issuecomment-852739729 : 

> "having another app (not in AutoKey) running which periodically keeps reading the clipboard and setting a timestamp whenever it changes. It could store that timestamp somewhere (e.g. by touching or editing a file) where your AutoKey script could retrieve it and know how old it is."

So that we could have something like: `clipboard.get_selection_timestamp()`

**Comment by josephj11** [comment]
I'm going to close this, but it does highlight the need for more documentation.

@ineuw It would be great if you could take *some* of the above :) and add it to our wiki - maybe under advanced scripting - with a title like "more than you ever wanted to know about clipboards" LOL. Also note that our API has separate calls for clipboard and selection for advanced users.

@vvasuki IMHO, this is a rare corner case that most people don't care about, so I would be against adding that functionality to AutoKey - especially because it would not be reliable. If I make a selection or copy something to the clipboard and go out to lunch, the clipboard is still perfectly valid even though the timestamp would be quite old.

This is analogous to the reason why no one uses file creation dates (and some systems don't even support them.) They sound useful, but upon examination, they don't tell you anything you can rely upon.

It would be very simple to write a bash script (or use any other language you prefer) which runs in the background updating a timestamp for the clipboard in a file. You could then write a Python function to retrieve that value from the file. You could even add that function to your AutoKey Modules directory so you could have one copy of it and import it into any other AutoKey script that needs it.

If you want to discuss this further, please post about it on our [support list](https://groups.google.com/forum/#!forum/autokey-users) because it's not issue related. I'd be happy to assist you there if you need help implementing it.

**Comment by Elliria** [comment]
This is a basic script that handles the clipboard without an exception, @vvasuki:
```python
# Fetch the clipboard content, giving it an empty string value if there's nothing on it:
try: 
    cb = clipboard.get_clipboard()
except: 
    cb = ""

# Print the contents of the variable:
keyboard.send_keys(cb)
```
